### ISSUE TITLE
Question about `verifytypes`

### ISSUE THREAD

**Issue by christopher-hesse** [microsoft/pyright#1301]
From https://github.com/microsoft/pyright/blob/master/docs/typed-libraries.md#examples-of-known-and-unknown-types:

```python
# Class with known type
class MyClass:
    height: float = 2.0

    def __init__(self, name: str, age: int):
        self.age: int = age

    @property
    def name(self) -> str:
        ...

# Class with partially unknown type
class MyClass:
    # Missing type annotation for class variable
    height = 2.0

    # Missing input parameter annotations
    def __init__(self, name, age):
        # Missing type annotation for instance variable
        self.age = age

    # Missing return type annotation
    @property
    def name(self):
        ...
```

In the second example `age` is an unknown type.  However if I make a class like:

```python
class MyClass:
    def __init__(self, name: str, age: int):
        self.age = age
```

`--verifytypes ` reports that the type is unknown for `age` (at least it does in a similar setup I have).  Why do member variables created directly from typed `__init__` args have unknown type?

**Comment by erictraut** [comment]
There are two reasons for this:

1. A member variable doesn't necessarily have the same type as the expression that's used to initialize it within a an `__init__` method. There are many cases where this would be a bad assumption on the part of a type checker. When pyright is inferring the type of a member variable (one that doesn't have an explicit type annotation), it consider all assignments of that member within all methods in the class. Consider, for example:

```python
class MyClass:
    def __init__(self, x: int):
        self.value = x

   def change_value(self, x: str):
       self.value = x
```
Pyright will infer that the type of `self.value` is `Union[int, str]` in this case.

2. Any symbol that is part of your library's public interface contract should have an explicit type declaration. Relying on a type checker's inference logic to determine a symbol's type is fraught with problems because inference rules are not defined in any PEP, and they differ somewhat from one type checker to another. For this reason, inference is not used when a library is marked as "py.typed". Instead, it is assumed that the library author has provided complete and accurate type declarations for all public symbols.

**Comment by christopher-hesse** [comment]
Got it, thanks for the explanation!
